src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java

```diff
@@ -149,4 +153,96 @@ void buildProductKeysQueryPredicate_WithSomeBlankProductKeys_ShouldBuildCorrectQ
         final QueryPredicate<Product> queryPredicate = service.buildProductKeysQueryPredicate(productKeys);
         assertThat(queryPredicate.toSphereQuery()).isEqualTo("key in (\"key1\", \"key2\")");
     }
+
+    @Test
+    void fetchMatchingProductsByKeys_WithBadGateWayExceptionAlways_ShouldFail() {
+        final FakeProductClient fakeProductClient = new FakeProductClient(new BadGatewayException());
+
+        initMockService(fakeProductClient);
+
+        final Set<String> keys =  new HashSet<>();
+        keys.add("productKey1");
+
+        CompletionStage<Set<Product>> future = service.fetchMatchingProductsByKeys(keys);
+
+        assertThat(errorExceptions).isEmpty();
+        assertThat(errorMessages).isEmpty();
+        assertThat(future.toCompletableFuture())
+                .failsWithin(60, TimeUnit.SECONDS)
+                .withThrowableOfType(ExecutionException.class)
+                .withCauseExactlyInstanceOf(BadGatewayException.class);
+    }
+
+    @Test
+    void fetchProduct_WithBadGatewayException_ShouldFail() {
+        final FakeProductClient fakeProductClient = new FakeProductClient(new BadGatewayException());
+
+        initMockService(fakeProductClient);
+
+        CompletionStage<Optional<Product>> future = service.fetchProduct("productKey1");
+
+        assertThat(errorExceptions).isEmpty();
+        assertThat(errorMessages).isEmpty();
+        assertThat(future.toCompletableFuture())
+                .failsWithin(60, TimeUnit.SECONDS)
```

Comment:
Why you need 60 seconds ?

Reply:
It is derived from a deprecated method `hasFailedWithThrowableThat()`
Javadoc suggests to use `failsWithin()` instead. It needs a timeout limit of service class async call, hence i put 60 seconds as default value here.

Reply:
OK, Javadocs of the hasFailedWithThrowableThat is suggesting 1 second.  Don't know where you got 60seconds default? Anyways you might skip my comment.

````
Although not 100% the same, consider using {@link #failsWithin(Duration)} or {@link #failsWithin(long, TimeUnit)} instead:`
   
  <pre><code class='java'> CompletableFuture future = new CompletableFuture();
  future.completeExceptionally(new RuntimeException("boom!"));
   
    assertThat(future).failsWithin(1, TimeUnit.SECONDS)
                      .withThrowableOfType(RuntimeException.class)
                      .withMessage("boom!");
````

Reply:
Choose it randomly.